### ISSUE TITLE
Unable to recover from "not currently connected and ResumeSubs not set" error

### ISSUE THREAD

**Issue by jan-bar** [eclipse/paho.mqtt.golang#666]
My business code is similar to the following. I subscribe to the topic in the `SetOnConnectHandler` callback. However, due to emqx restart and other reasons, a certain state cannot be restored when reconnecting.

```go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

func main() {
	opts := mqtt.NewClientOptions().AddBroker("tcp://127.0.0.1:8883")
	opts.SetOnConnectHandler(func(client mqtt.Client) {
		for {
			tk := client.Subscribe("topic", 0, func(client mqtt.Client, message mqtt.Message) {
				// do something
			})
			if tk.WaitTimeout(time.Minute) && tk.Error() == nil {
				break // if the Subscribe is successful, go to the next step.
			}
			// Subscribe failed print error and try again
			fmt.Println(tk.Error())
			time.Sleep(time.Minute)
		}
	})

	c := mqtt.NewClient(opts)
	if token := c.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	sig := make(chan os.Signal, 1)
	signal.Notify(sig, os.Interrupt)
	<-sig

	c.Disconnect(250)
}
```

The service has not returned to normal for 2 days after the problem occurred. The following is my error log. I tried rebooting and it ran normally.
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/f60bdbac-456c-4717-8589-d658e1390265)

I found that the source code keeps returning errors as shown below.
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/62733d52-7be1-45bf-9ae4-94d0380e112e)

I am now trying to reproduce the problem, and I have restarted emqx many times but it has not reappeared. I suspect that there should be other factors causing this problem, such as the pressure on the production environment's CPU and memory.

At present, I don't have any good solution. I can only write a subscription timeout to trigger panic and automatically restart. Of course, this is not a good solution.

**Comment by MattBrittan** [comment]
Unfortunately there is nothing in your question that would provide any info on why you were not connecting and code that is "similar to the following" is of limited use (unless you can confirm that the code you provided can replicate the issue; the library has lots of options so detail is important).

Based on the number of times the error is repeating in the logs I'd guess that your `for` loop was running in a significant number of go-routines (`OnConnectHandler` is called in a go-routine). What I would guess is happening is:

 * Connection to broker comes up
 * `OnConnectHandler` called in go routine
 * Connection to broker drops
 * Your `OnConnectHandler` calls `Subscribe`

A classic cause of this behaviour would be another client running with the same client ID (each time you connect the broker kills the other connection, that client then reconnects meaning the broker kills the other connection, so you get an infinite loop).

Personally I'd suggest removing the `for` loop here. If `Subscribe` fails then retrying it is unlikely to really help (either the connection has dropped, in which case `Subscribe` will be retried when it comes up again, or the broker rejected the request and is unlikely to change it's decision).

**Comment by jan-bar** [comment]
@MattBrittan Sorry, I completed all the configuration below.   I'm not even sure now whether subscribing to a topic in a SetOnConnectHandler callback is common practice.

In addition, when reconnecting, we know from the log that `IsConnected() == true`,`IsConnectionOpen() == false`, which corresponds to the error path when subscribing.  I want to know why this state persists. I tried restarting emqx manually but it didn't happen. By reading the source code, I haven't found the trigger mechanism yet.

Client id conflicts are impossible, I randomize the client id every time.

Because re-subscription is required when reconnecting, what should be done if the internal subscription of SetOnConnectHandler fails at this time? How to actively trigger mqtt reconnection?

```go
package main

import (
	"crypto/tls"
	"fmt"
	"os"
	"os/signal"
	"time"

	mqtt "github.com/eclipse/paho.mqtt.golang"
)

func main() {
	opts := mqtt.NewClientOptions().AddBroker("tcp://127.0.0.1:8883")
	opts.SetConnectRetry(true).SetConnectRetryInterval(time.Second * 5)
	opts.SetConnectTimeout(time.Second * 10).SetAutoReconnect(true)
	opts.SetKeepAlive(time.Second * 10).SetCleanSession(true)
	opts.SetTLSConfig(&tls.Config{InsecureSkipVerify: true})
	opts.SetOrderMatters(true)
        opts.SetConnectionLostHandler(func(c MQTT.Client, err error) {
		log.Errorf("MQTT Disconnected,IsConnected:%v,IsConnectionOpen:%v,%+v",
			c.IsConnected(), c.IsConnectionOpen(), err)
	})
	opts.SetOnConnectHandler(func(client mqtt.Client) {
		for {
			tk := client.Subscribe("topic", 0, func(client mqtt.Client, message mqtt.Message) {
				// do something
			})
			if tk.WaitTimeout(time.Minute) && tk.Error() == nil {
				break // if the Subscribe is successful, go to the next step.
			}
			// Subscribe failed print error and try again
			fmt.Println(tk.Error())
			time.Sleep(time.Minute)
		}
	})

	c := mqtt.NewClient(opts)
	if token := c.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	sig := make(chan os.Signal, 1)
	signal.Notify(sig, os.Interrupt)
	<-sig

	c.Disconnect(250)
}
```

**Comment by MattBrittan** [comment]
> we know from the log that IsConnected() == true,IsConnectionOpen() == false

* `IsConnected()` will always be `true` (unless you call `Disconnect`) - as per the docs "onnected means that the connection is up now OR it will be established/reestablished automatically when possible"
* `IsConnectionOpen()` is the true status of the connection (as the client knows it at the time).

>Because re-subscription is required when reconnecting, what should be done if the internal subscription of SetOnConnectHandler fails at this time? How to actively trigger mqtt reconnection?

Possible errors from Subscribe are (not 100% sure I got them all):
* "not currently connected and ResumeSubs not set"
* "reconnecting state and cleansession is true"
* "connection lost before Subscribe completed"
* "no message IDs available"
* "subscribe was broken by timeout"

The bottom two would be the only ones that might, possibly, be worth retrying (but the likelihood of these being returned AND the connection remaining up seems minimal). Note that what you really should be checking is the `tk.(*SubscribeToken).Result()` (to see if the broker rejected any of the requests). At a minimum I'd be adding a check on `IsConnectionOpen()` into your loop (`return` if `false` is returned as the callback will be called again when the connection comes up).

The challenge for your handler is that it's called in a goroutine so should not make assumptions about the connection status (which may have changed since it was started). Terminating the connection is problematic because the library will probably be attempting to re-establish it (so it might reconnect successfully only to have your handler drop the connection again).

>What I'm curious about is why the paho.mqtt.golang library cannot automatically return to normal when this state occurs.

Unfortunately there is still not enough information to ascertain this (you have reported on some of the symptoms of the connection dropping but nothing on the cause). Taking a look at the logs (either client or broker) is probably the first step. You might wish to run your sample app somewhere (connecting to the same broker) and see if it has the same issue (it's quite possible that it will work 100% and the issue is elsewhere).

**Comment by jan-bar** [comment]
@MattBrittan 

Thanks for the reply, I will try to reproduce the problem in my environment as you suggested. This may indeed be a problem with my usage, I will try to troubleshoot and report the results here.

**Comment by jan-bar** [comment]
@MattBrittan 

By looking at the logs from the last time my program failed to recover, I found the critical error message `"pingresp not received, disconnecting"` .  At that time, the entire Linux system should be restarted, and the emqx service started later than my program.

![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/6b53cc21-396d-4152-920d-6516cb5081cb)

I guess it's a problem with this part of the code
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/89d1556e-c669-456a-8cc0-17520e8f28df)

I guess the code runs as follows. Because keepalive is a coroutine, it is entirely possible to cause the following situation. This is also the reason why it is difficult for me to reproduce by restarting emqx.
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/8274cb77-bdf6-4a42-95ff-304698f27616)

I add the reconnect parameter to `startCommsWorkers` to ensure that my guess logic will only enter when reconnecting. And artificially create the process I guessed when reconnecting, and simulate the coroutine to execute logic successively through time.Sleep
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/ac85f640-64cf-43e7-a047-78aa8c737c99)

By adding parameters to ensure normal connection and entering the original process, the process I added will only be entered when the server is reconnected.
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/072fd16f-8a30-47e9-bbe0-90c1daf290af)

First run normally and receive a message through subscription, then restart the emqx server, and finally I reproduced the problem.
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/042b7177-9e50-4b1a-8b13-09163b37b2cb)

By reading the source code and analyzing the process, this should be the cause. I'm not sure if this is a bug, I'm still thinking about how to fix it. Do you have any good suggestions?

**Comment by MattBrittan** [comment]
Sorry; I'm not sure what you believe your demonstration code is showing. If I duplicate this (with full logging) I get: 

```
[ERROR] [client]   dial tcp 127.0.0.1:1883: connectex: No connection could be made because the target machine actively refused it.
[WARN]  [client]   failed to connect to broker, trying next
[DEBUG] [client]   Reconnect failed, slept for 2 seconds: network Error : dial tcp 127.0.0.1:1883: connectex: No connection could be made because the target machine actively refused it.
[DEBUG] [client]   about to write new connect msg
[DEBUG] [client]   socket connected to broker
[DEBUG] [client]   Using MQTT 3.1.1 protocol
[DEBUG] [net]      connect started
[DEBUG] [net]      received connack
[DEBUG] [client]   startCommsWorkers called
[DEBUG] [client]   client is connected/reconnected
[DEBUG] [pinger]   keepalive starting
[DEBUG] [net]      incoming started
[DEBUG] [net]      startIncomingComms started
[DEBUG] [net]      outgoing started
[DEBUG] [net]      startComms started
[DEBUG] [net]      logic waiting for msg on ibound
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [client]   startCommsWorkers done
[DEBUG] [store]    enter Resume
[DEBUG] [store]    exit resume
[DEBUG] [net]      startIncomingComms: inboundFromStore complete
[DEBUG] [net]      logic waiting for msg on ibound
[DEBUG] [client]   internalConnLost called
[DEBUG] [client]   stopCommsWorkers called
[DEBUG] [pinger]   keepalive stopped
[DEBUG] [net]      incoming complete
[DEBUG] [client]   internalConnLost waiting on workers
[DEBUG] [client]   stopCommsWorkers waiting for workers
[DEBUG] [net]      startIncomingComms: ibound complete
[DEBUG] [net]      startIncomingComms goroutine complete
[DEBUG] [router]   matchAndDispatch exiting
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [net]      outgoing comms stopping
[DEBUG] [client]   startCommsWorkers output redirector finished
[DEBUG] [client]   stopCommsWorkers waiting for comms
[DEBUG] [net]      startComms closing outError
[DEBUG] [client]   incoming comms goroutine done
[DEBUG] [client]   stopCommsWorkers done
[DEBUG] [client]   internalConnLost workers stopped
[DEBUG] [msgids]   cleaned up subs
[DEBUG] [client]   internalConnLost complete
2024/01/14 09:21:22 MQTT Disconnected,IsConnected:true,IsConnectionOpen:false,test crash
[DEBUG] [client]   enter reconnect
[DEBUG] [client]   enter Subscribe
not currently connected and ResumeSubs not set
[DEBUG] [client]   Detect continual connection lost after reconnect, slept for 1 seconds
[DEBUG] [client]   about to write new connect msg
[DEBUG] [client]   socket connected to broker
[DEBUG] [client]   Using MQTT 3.1.1 protocol
[DEBUG] [net]      connect started
[DEBUG] [net]      received connack
[DEBUG] [client]   startCommsWorkers called
[DEBUG] [client]   client is connected/reconnected
[DEBUG] [pinger]   keepalive starting
[DEBUG] [net]      incoming started
[DEBUG] [net]      startIncomingComms started
[DEBUG] [net]      outgoing started
[DEBUG] [net]      startComms started
[DEBUG] [client]   startCommsWorkers done
[DEBUG] [store]    enter Resume
[DEBUG] [net]      logic waiting for msg on ibound
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [store]    exit resume
[DEBUG] [net]      startIncomingComms: inboundFromStore complete
[DEBUG] [net]      logic waiting for msg on ibound
[DEBUG] [client]   internalConnLost called
[DEBUG] [client]   stopCommsWorkers called
[DEBUG] [pinger]   keepalive stopped
[DEBUG] [net]      incoming complete
[DEBUG] [client]   internalConnLost waiting on workers
[DEBUG] [net]      startIncomingComms: ibound complete
[DEBUG] [net]      startIncomingComms goroutine complete
[DEBUG] [client]   stopCommsWorkers waiting for workers
[DEBUG] [router]   matchAndDispatch exiting
[DEBUG] [client]   startCommsWorkers output redirector finished
[DEBUG] [client]   stopCommsWorkers waiting for comms
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [net]      outgoing waiting for an outbound message
[DEBUG] [net]      outgoing comms stopping
[DEBUG] [net]      startComms closing outError
[DEBUG] [client]   incoming comms goroutine done
[DEBUG] [client]   stopCommsWorkers done
[DEBUG] [client]   internalConnLost workers stopped
[DEBUG] [msgids]   cleaned up subs
[DEBUG] [client]   internalConnLost complete
[DEBUG] [client]   enter reconnect
2024/01/14 09:21:24 MQTT Disconnected,IsConnected:true,IsConnectionOpen:false,test crash
[DEBUG] [client]   enter Subscribe
```

This is as-expected; you end up with a loop:
* Client connects
* go routine 1 starts (waits a second then triggers disconnect)
* go routine 2 starts (waits two second then calls OnConnect)
* routine 1 triggers reconnection (I used `c.internalConnLost(errors.New("test crash"))`).
* routine 2 triggers OnConnect that (probably) starts printing errors because connection is down
* Client begins another reconnection attempt

So when I run this if just loops reconnecting, losing connection, reconnecting etc (as would be expected). If you believe this behaviour is at issue then please run your example with full logging enabled i.e.

```
	mqtt.ERROR = log.New(os.Stdout, "[ERROR] ", 0)
	mqtt.CRITICAL = log.New(os.Stdout, "[CRIT] ", 0)
	mqtt.WARN = log.New(os.Stdout, "[WARN]  ", 0)
	mqtt.DEBUG = log.New(os.Stdout, "[DEBUG] ", 0)
```

In addition please consider storing your replication code somewhere public (i.e. github). I can copy from the screenshots but its a little slow, and I cannot be sure I'm doing exactly what you are (note that I am testing with Mosquitto but that should not make a difference here).

**Comment by jan-bar** [comment]
@MattBrittan I made a test code that must appear.  You need to execute `go build -mod=vendor` to compile the code.

[test_mqtt.zip](https://github.com/eclipse/paho.mqtt.golang/files/13929987/test_mqtt.zip)

In the code in question, there is an infinite loop, which means that the loop in the picture below will not exit, which will cause the current bug.  I've tried exiting `SetOnConnectHandler` callback after the problem occurs and having the code not enter an exception state. At this point the program will automatically return to normal.
![image](https://github.com/eclipse/paho.mqtt.golang/assets/14202095/907c7efa-ea2d-4cd7-a13c-03e6e91c9151)

I would like to ask if this test can show that I don't need to handle the subscription failure error in `SetOnConnectHandler` because `paho.mqtt.golang` will automatically restore the state and then continue to call the callback of `SetOnConnectHandler`

**Comment by MattBrittan** [comment]
Sorry - I'm not sure what you are intending this test to prove.

Following loss of connection it will forcefully drop the connection 20 times before allowing it to proceed; this seems to work as expected:

```
client.go:829: [client]   enter Subscribe
not currently connected and ResumeSubs not set
------------------------
client.go:315: [client]   Detect continual connection lost after reconnect, slept for 32 seconds
client.go:369: [client]   about to write new connect msg
client.go:394: [client]   socket connected to broker
net.go:62: [client]   Using MQTT 3.1.1 protocol
net.go:81: [net]      connect started
net.go:100: [net]      received connack
client.go:579: [client]   startCommsWorkers called
client.go:618: [client]   client is connected/reconnected
net.go:120: [net]      incoming started
net.go:163: [net]      startIncomingComms started
net.go:262: [net]      outgoing started
SetOnConnectHandler,IsConnected:true,IsConnectionOpen:true
net.go:395: [net]      startComms started
```

Again this seems to be operating as designed (it recovers from the loss of connection).


>show that I don't need to handle the subscription failure error in SetOnConnectHandler because paho.mqtt.golang will automatically restore the state and then continue to call the callback of SetOnConnectHandler

`paho.mqtt.golang` will not restore the subscription (if you used `SetCleanSession(false)` then the broker should remember it). However every time you connect the subscription will be retried so, as I suggested before, there is no point looping in `OnConnectHandler` if `!IsConnectionOpen()` (because the callback will be called again when the connection comes up).

**Comment by jan-bar** [comment]
@MattBrittan Hmm, this test matches the cause of my problem. I will also remove the loop in SetOnConnectHandler. If the subscription fails due to connection, it can eventually return to normal automatically. Thank you very much for your answer.

**Comment by MattBrittan** [comment]
No worries. If you want to really ensure resilience one approach would be to add a timeout based upon when you expect to receive messages (e.g. if no message received in 10 minutes then restart). In reality I've had a lot of instances running for years that ignore the return from `Subscribe` and have never had an issue (if it fails then it's almost certain that the network link will drop too).

**Comment by klimov-ankr** [comment]
I do like that, is that right?
```go
timer := time.NewTimer(3 * time.Minute)
	for !nodeClient.MQTTClient.IsConnectionOpen() {
		if !nodeClient.MQTTClient.IsConnected() {
			if err := nodeClient.Connect(context.Background()); err != nil {
				return nil, fmt.Errorf("failed to connect to broker: %s", err)
			}
			continue
		}
		select {
		case <-timer.C:
			return nil, fmt.Errorf("failed to connect within 3 minutes")
		case <-time.After(5 * time.Second):
			continue
		case <-ctx.Done():
			return nil, nil
		}
	}
	```

**Comment by MattBrittan** [comment]
> I do like that, is that correct?

Please dont add comments to old, closed issues. Raise a new issue, or preferably, as this is a question, ask on Stackoverflow (chances are I'll answer there, but there are others watching whereas here its unlikely anyone else will see your question). Note that i would recommended just using the inbuilt reconnection functionality (this is difficult to get right so dont do it yourself unless you have to). As it stands you code is confusing (once `token.Wait` returns the connection attempt is complete or an error has occured).
